File: macos/Tests/VoiceKeyCoreTests/VocabularyEditorTests.swift
import Testing
// Foundation deliberately not imported here — see FoundationExport.swift.
@testable import VoiceKeyCore

@Suite struct VocabularyEditorTests {

    // MARK: - Loading

    @Test func replacementRowsAreSortedBySpokenPhrase() {
        let editor = makeEditor(replacements: ["zebra": "Z", "alpha": "A", "middle": "M"])

        #expect(editor.replacements.map(\.spoken) == ["alpha", "middle", "zebra"])
        #expect(editor.replacements.map(\.written) == ["A", "M", "Z"])
    }

    @Test func termsKeepTheOrderTheyWereSavedIn() {
        let editor = makeEditor(terms: ["Kubernetes", "VoiceKey", "Herglotz"])

        #expect(editor.terms == ["Kubernetes", "VoiceKey", "Herglotz"])
    }

    @Test func anEmptyDictionaryOpensWithNoRows() {
        let editor = VocabularyEditor()

        #expect(editor.replacements.isEmpty)
        #expect(editor.terms.isEmpty)
    }

    // MARK: - Saving

    @Test func editedRowsRoundTripBackIntoTheDictionary() {
        let original = VocabularyDictionary(terms: ["WhisperKit"],
                                            replacements: ["get hub": "GitHub"])

        #expect(VocabularyEditor(dictionary: original).dictionary == original)
    }

    @Test func aHalfTypedRowIsNotSavedAsARule() {
        let editor = makeEditor(replacements: ["get hub": "GitHub"])
            .addingReplacement()
            .settingSpoken("jason", at: 1)

        #expect(editor.replacements.count == 2, "the half-typed row stays visible while editing")
        #expect(editor.dictionary.replacements == ["get hub": "GitHub"])
    }

    @Test func blankTermsAreNotSaved() {
        let editor = makeEditor(terms: ["VoiceKey"]).addingTerm()

        #expect(editor.terms.count == 2)
        #expect(editor.dictionary.terms == ["VoiceKey"])
    }

    @Test func surroundingWhitespaceIsTrimmedOnSave() {
        let editor = VocabularyEditor()
            .addingReplacement()
            .settingSpoken("  get hub  ", at: 0)
            .settingWritten("  GitHub  ", at: 0)
            .addingTerm()
            .settingTerm("  WhisperKit  ", at: 0)

        #expect(editor.dictionary.replacements == ["get hub": "GitHub"])
        #expect(editor.dictionary.terms == ["WhisperKit"])
    }

    @Test func theFirstOfTwoRowsWithTheSameSpokenPhraseWins() {
        let editor = VocabularyEditor()
            .addingReplacement()
            .settingSpoken("jason", at: 0)
            .settingWritten("JSON", at: 0)
            .addingReplacement()
            .settingSpoken("Jason", at: 1)
            .settingWritten("Jason Bourne", at: 1)

        #expect(editor.dictionary.replacements == ["jason": "JSON"],
                "the row shown first is the one that takes effect")
    }

    @Test func duplicateSpokenPhrasesAreFlaggedForTheUserCaseInsensitively() {
        let editor = VocabularyEditor()
            .addingReplacement()
            .settingSpoken("jason", at: 0)
            .settingWritten("JSON", at: 0)
            .addingReplacement()
            .settingSpoken(" JASON ", at: 1)
            .settingWritten("Jason Bourne", at: 1)
            .addingReplacement()
            .settingSpoken("get hub", at: 2)
            .settingWritten("GitHub", at: 2)

        #expect(editor.ignoredReplacementRows == [1],
                "only the shadowed row is flagged, not the one that takes effect")
    }

    @Test func emptyRowsAreNotFlaggedAsDuplicatesOfEachOther() {
        let editor = VocabularyEditor().addingReplacement().addingReplacement()

        #expect(editor.ignoredReplacementRows.isEmpty)
    }

    // MARK: - Editing

    @Test func addingAReplacementAppendsAnEmptyRowAtTheEnd() {
        let editor = makeEditor(replacements: ["get hub": "GitHub"]).addingReplacement()

        #expect(editor.replacements.count == 2)
        #expect(editor.replacements.last == VocabularyEditor.Replacement(spoken: "", written: ""))
    }

    @Test func editingOneRowLeavesTheOthersAlone() {
        let editor = makeEditor(replacements: ["alpha": "A", "beta": "B"])
            .settingWritten("Bravo", at: 1)

        #expect(editor.dictionary.replacements == ["alpha": "A", "beta": "Bravo"])
    }

    @Test func removingRowsDropsExactlyTheSelectedOnes() {
        let editor = makeEditor(replacements: ["alpha": "A", "beta": "B", "gamma": "G"])
            .removingReplacements(at: [0, 2])

        #expect(editor.dictionary.replacements == ["beta": "B"])
    }

    @Test func removingTermsDropsExactlyTheSelectedOnes() {
        let editor = makeEditor(terms: ["one", "two", "three"]).removingTerms(at: [1])

        #expect(editor.terms == ["one", "three"])
    }

    @Test func termsCanBeAddedAndEdited() {
        let editor = VocabularyEditor().addingTerm().settingTerm("Kubernetes", at: 0)

        #expect(editor.dictionary.terms == ["Kubernetes"])
    }

    @Test func editsToRowsThatDoNotExistAreIgnored() {
        let editor = makeEditor(replacements: ["alpha": "A"], terms: ["one"])

        #expect(editor.settingSpoken("x", at: 7) == editor)
        #expect(editor.settingWritten("x", at: -1) == editor)
        #expect(editor.settingTerm("x", at: 3) == editor)
        #expect(editor.removingReplacements(at: [9]) == editor)
        #expect(editor.removingTerms(at: [9]) == editor)
    }

    // MARK: - Helpers

    private func makeEditor(
        replacements: [String: String] = [:],
        terms: [String] = []
    ) -> VocabularyEditor {
        VocabularyEditor(dictionary: VocabularyDictionary(terms: terms, replacements: replacements))
    }
}

File: windows/VoiceKey.Core.Tests/VocabularyEditorTests.cs
namespace VoiceKey.Core.Tests;

public class VocabularyEditorTests
{
    // MARK: - Loading

    [Fact]
    public void ReplacementRowsAreSortedBySpokenPhrase()
    {
        var editor = MakeEditor(replacements: new()
        {
            ["zebra"] = "Z",
            ["alpha"] = "A",
            ["middle"] = "M",
        });

        Assert.Equal(["alpha", "middle", "zebra"], editor.Replacements.Select(row => row.Spoken));
        Assert.Equal(["A", "M", "Z"], editor.Replacements.Select(row => row.Written));
    }

    [Fact]
    public void TermsKeepTheOrderTheyWereSavedIn()
    {
        var editor = MakeEditor(terms: ["Kubernetes", "VoiceKey", "Herglotz"]);

        Assert.Equal(["Kubernetes", "VoiceKey", "Herglotz"], editor.Terms);
    }

    [Fact]
    public void AnEmptyDictionaryOpensWithNoRows()
    {
        var editor = new VocabularyEditor();

        Assert.Empty(editor.Replacements);
        Assert.Empty(editor.Terms);
    }

    // MARK: - Saving

    [Fact]
    public void EditedRowsRoundTripBackIntoTheDictionary()
    {
        var original = new VocabularyDictionary(
            terms: ["Whisper"],
            replacements: new Dictionary<string, string> { ["get hub"] = "GitHub" });

        Assert.Equal(original, new VocabularyEditor(original).Dictionary);
    }

    [Fact]
    public void AHalfTypedRowIsNotSavedAsARule()
    {
        var editor = MakeEditor(replacements: new() { ["get hub"] = "GitHub" })
            .AddingReplacement()
            .SettingSpoken("jason", 1);

        Assert.Equal(2, editor.Replacements.Count);
        Assert.Equal(Rules(("get hub", "GitHub")), editor.Dictionary);
    }

    [Fact]
    public void BlankTermsAreNotSaved()
    {
        var editor = MakeEditor(terms: ["VoiceKey"]).AddingTerm();

        Assert.Equal(2, editor.Terms.Count);
        Assert.Equal(["VoiceKey"], editor.Dictionary.Terms);
    }

    [Fact]
    public void SurroundingWhitespaceIsTrimmedOnSave()
    {
        var editor = new VocabularyEditor()
            .AddingReplacement()
            .SettingSpoken("  get hub  ", 0)
            .SettingWritten("  GitHub  ", 0)
            .AddingTerm()
            .SettingTerm("  Whisper  ", 0);

        Assert.Equal(Rules(("get hub", "GitHub")).Replacements, editor.Dictionary.Replacements);
        Assert.Equal(["Whisper"], editor.Dictionary.Terms);
    }

    [Fact]
    public void TheFirstOfTwoRowsWithTheSameSpokenPhraseWins()
    {
        var editor = new VocabularyEditor()
            .AddingReplacement().SettingSpoken("jason", 0).SettingWritten("JSON", 0)
            .AddingReplacement().SettingSpoken("Jason", 1).SettingWritten("Jason Bourne", 1);

        Assert.Equal(Rules(("jason", "JSON")), editor.Dictionary);
    }

    [Fact]
    public void DuplicateSpokenPhrasesAreFlaggedForTheUserCaseInsensitively()
    {
        var editor = new VocabularyEditor()
            .AddingReplacement().SettingSpoken("jason", 0).SettingWritten("JSON", 0)
            .AddingReplacement().SettingSpoken(" JASON ", 1).SettingWritten("Jason Bourne", 1)
            .AddingReplacement().SettingSpoken("get hub", 2).SettingWritten("GitHub", 2);

        Assert.Equal([1], editor.IgnoredReplacementRows.Order());
    }

    [Fact]
    public void EmptyRowsAreNotFlaggedAsDuplicatesOfEachOther()
    {
        var editor = new VocabularyEditor().AddingReplacement().AddingReplacement();

        Assert.Empty(editor.IgnoredReplacementRows);
    }

    // MARK: - Editing

    [Fact]
    public void AddingAReplacementAppendsAnEmptyRowAtTheEnd()
    {
        var editor = MakeEditor(replacements: new() { ["get hub"] = "GitHub" }).AddingReplacement();

        Assert.Equal(2, editor.Replacements.Count);
        Assert.Equal(new Replacement("", ""), editor.Replacements[^1]);
    }

    [Fact]
    public void EditingOneRowLeavesTheOthersAlone()
    {
        var editor = MakeEditor(replacements: new() { ["alpha"] = "A", ["beta"] = "B" })
            .SettingWritten("Bravo", 1);

        Assert.Equal(Rules(("alpha", "A"), ("beta", "Bravo")), editor.Dictionary);
    }

    [Fact]
    public void RemovingRowsDropsExactlyTheSelectedOnes()
    {
        var editor = MakeEditor(replacements: new()
        {
            ["alpha"] = "A",
            ["beta"] = "B",
            ["gamma"] = "G",
        }).RemovingReplacements([0, 2]);

        Assert.Equal(Rules(("beta", "B")), editor.Dictionary);
    }

    [Fact]
    public void RemovingTermsDropsExactlyTheSelectedOnes()
    {
        var editor = MakeEditor(terms: ["one", "two", "three"]).RemovingTerms([1]);

        Assert.Equal(["one", "three"], editor.Terms);
    }

    [Fact]
    public void TermsCanBeAddedAndEdited()
    {
        var editor = new VocabularyEditor().AddingTerm().SettingTerm("Kubernetes", 0);

        Assert.Equal(["Kubernetes"], editor.Dictionary.Terms);
    }

    [Fact]
    public void EditsToRowsThatDoNotExistAreIgnored()
    {
        var editor = MakeEditor(replacements: new() { ["alpha"] = "A" }, terms: ["one"]);

        Assert.Equal(editor, editor.SettingSpoken("x", 7));
        Assert.Equal(editor, editor.SettingWritten("x", -1));
        Assert.Equal(editor, editor.SettingTerm("x", 3));
        Assert.Equal(editor, editor.RemovingReplacements([9]));
        Assert.Equal(editor, editor.RemovingTerms([9]));
    }

    // MARK: - Helpers

    private static VocabularyEditor MakeEditor(
        Dictionary<string, string>? replacements = null,
        IReadOnlyList<string>? terms = null) =>
        new(new VocabularyDictionary(terms ?? [], replacements ?? []));

    private static VocabularyDictionary Rules(params (string Spoken, string Written)[] rules) =>
        new(replacements: rules.ToDictionary(rule => rule.Spoken, rule => rule.Written));
}
